src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1186,6 +1212,45 @@ impl LinkCollector<'_, '_> {
         }
     }
 
+    fn resolve_with_disambiguator_cached(
+        &mut self,
+        key: CacheKey,
+        item: &Item,
+        dox: &str,
+        ori_link: &str,
+        link_range: Option<Range<usize>>,
+    ) -> Option<(Res, Option<String>)> {
+        // Try to look up both the result and the corresponding side channel value
+        if let Some(ref cached) = self.visited_links.get(&key) {
+            self.kind_side_channel.set(cached.side_channel.clone());
+            Some(cached.res.clone())
+        } else {
+            match self.resolve_with_disambiguator(
+                key.dis,
+                item,
+                dox,
+                &key.path_str,
+                key.module_id,
+                key.extra_fragment.clone(),
+                ori_link,
+                link_range,
+            ) {
+                Some(res) => {
+                    // Store result for the actual namespace
+                    self.visited_links.insert(
+                        key,
+                        CachedLink {
+                            res: res.clone(),
+                            side_channel: self.kind_side_channel.clone().into_inner(),
+                        },
+                    );
+                    Some(res)
+                }
+                _ => None,
+            }
+        }
+    }
+
     /// After parsing the disambiguator, resolve the main part of the link.
     // FIXME(jynelson): wow this is just so much
     fn resolve_with_disambiguator(
```

Comment:
Can you update this to take a `CachedKey` so there are fewer arguments? Since we have it anyway it seems like a good time for a cleanup.

Reply:
That feels super weird as I personally don't think we should mix parts of the cache related code into the "uncached" function.

Reply:
Hmm, I think they match naturally. `resolve_with_disambiguator` has two types of arguments:
1. Arguments used for resolving the link, like `path_str` and `module_id`
2. Arguments used for diagnostics, like `item` and `link_range`.

I think it's nice to separate them into structs a little, so it's clear which are which.

Reply:
I'll see what I can do, maybe CacheKey could be renamed into something like `ResolutionInfo` or something similar, if we're going to introduce `DiagnosticInfo` as well. That rename would make this change less awkward for me brain.